### PR TITLE
feat(components): [popper] focus trap and a11y

### PR DESCRIPTION
**Pull request by opengraphica**
/label Components:[popper]

Please make sure these boxes are checked before submitting your PR, thank you!

- [x] Make sure you follow contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
- [x] Make sure you are merging your commits to `dev` branch.
- [x] Add some descriptions and refer to relative issues for your PR.

Primary goal of this PR is to add focus trap to popper, and fix any obvious bugs in existing components using popper where focus trapping is involved, such as anything that allows an unexpected escape of the trap, or "esc" key not working correctly when nested inside of a dialog.

Many of these components do not currently have keyboard-accessible poppers, these will be tackled in detail individually in future PRs.

[focus-trap]
- Changed to auto-add keydown listener based on forwardRef
- Added new emit `focusout-prevented` so parent component can choose to release trap when tab out
- Added new emit `release-requested` which fires when esc is pressed only when the trap is active

[popper]
- Added focus trap
- New prop "focus-on-show", which will auto focus on the popper content whenever it shows. Usually this isn't desirable.
- New prop "trapping". When true, keyboard tab will be trapped in the popper. When false, tabbing outside the popper will close the focus layer and focus on the element before focus was brought in to the poopper.

[tooltip]
- Closes automatically when the focus trap layer closes.

[autocomplete]
- Made escape key close popper

[cascader]
- Ensure escape key doesn't also close dialog when popper open

[color-picker]
- No change

[dropdown]
- Removed focus trap inside this component as it steals focus trap from dialog. This trap didn't seem to be working anyways? Future work.

[menu]
- No change

[message-box]
- Listen to `release-requested` emit on focus trap to handle esc key close

[popover]
- No change

[pop-confirm]
- No change

[select]
- Escape key in input only prevent default if popper is visible

[time-select]
- No change

[tree-select]
- No change

### CODE REVIEW

**Comment by opengraphica** [comment]
`/label Components:[popper]`

File: packages/components/focus-trap/src/focus-trap.vue [hunk 1]
```diff
@@ -108,31 +126,59 @@ export default defineComponent({
...
+        forwardRef.addEventListener('focusin', onFocusIn)
+        forwardRef.addEventListener('focusout', onFocusOut)
+      }
+      if (oldForwardRef) {
+        oldForwardRef.removeEventListener('keydown', onKeydown)
+        oldForwardRef.removeEventListener('focusin', onFocusIn)
+        oldForwardRef.removeEventListener('focusout', onFocusOut)
+      }
+    })
+
```
Comment (JeremyWuuuuu) [inline packages/components/focus-trap/src/focus-trap.vue:141]: **Suggestion:**

```suggestion
    watch(forwardRef, (forwardRef, oldForwardRef) => {
      if (forwardRef) {
        forwardRef.addEventListener('keydown', onKeydown)
        forwardRef.addEventListener('focusin', onFocusIn)
        forwardRef.addEventListener('focusout', onFocusOut)
      }
      if (oldForwardRef) {
        oldForwardRef.removeEventListener('keydown', onKeydown)
        oldForwardRef.removeEventListener('focusin', onFocusIn)
        oldForwardRef.removeEventListener('focusout', onFocusOut)
      }
    })

```

File: packages/components/popper/src/content.ts [hunk 2]
```diff
@@ -55,6 +55,14 @@ export const usePopperContentProps = buildProps({
...
   },
   pure: Boolean,
+  focusOnShow: {
+    type: Boolean,
+    default: false,
+  },
+  trapping: {
+    type: Boolean,
+    default: false,
+  },
```
Comment (JeremyWuuuuu) [inline packages/components/popper/src/content.ts:65]:
```suggestion
  focusOnShow: Boolean,
  trapping: Boolean,
```

Much simpler.

File: packages/components/popper/src/content.vue [hunk 3]
```diff
@@ -10,20 +10,33 @@
     @mouseenter="(e) => $emit('mouseenter', e)"
     @mouseleave="(e) => $emit('mouseleave', e)"
   >
-    <slot />
+    <el-focus-trap
+      :trapped="trapped"
+      :trap-on-focus-in="true"
```
Comment (JeremyWuuuuu) [inline packages/components/popper/src/content.vue:15]:
```suggestion
      trap-on-focus-in
```

File: packages/hooks/use-escape-keydown/index.ts [hunk 4]
```diff
@@ -1,19 +1,32 @@
...
     }
   }
+
   onMounted(() => {
-    on(document, 'keydown', cachedHandler)
+    if (registeredEscapeHandlers.length === 0) {
+      document.addEventListener('keydown', cachedHandler)
+    }
+    if (isClient) registeredEscapeHandlers.push(handler)
   })
```
Comment (JeremyWuuuuu) [inline packages/hooks/use-escape-keydown/index.ts:22]: IMHO, `onMounted` will only get invoked on client side, so I do not think it is necessary to check that here.

File: packages/hooks/use-escape-keydown/index.ts [hunk 5]
```diff
@@ -1,19 +1,32 @@
...
 
   onBeforeUnmount(() => {
-    off(document, 'keydown', cachedHandler)
+    registeredEscapeHandlers = registeredEscapeHandlers.filter(
+      (registeredHandler) => registeredHandler !== handler
+    )
+    if (registeredEscapeHandlers.length === 0) {
+      if (isClient) document.removeEventListener('keydown', cachedHandler)
```
Comment (JeremyWuuuuu) [inline packages/hooks/use-escape-keydown/index.ts:29]: Maybe the checking here is legit.

**Review by jw-foss — COMMENTED** [review]
@opengraphica I have left a few comments for you to consider and sorry for the late review, back to you 🏓

**Comment by JeremyWuuuuu** [comment]
@opengraphica I think I will merge this one and resolve the comments on my own. We are preparing for a new release

**Review by jw-foss — APPROVED** [review]

File: packages/hooks/use-escape-keydown/index.ts (hunk 4 above)
Comment (Giwayume) [reply packages/hooks/use-escape-keydown/index.ts:22]: Yeah I just moved that isClient check since it was there in the previous code

**Pull request merged**